AX: live region announcements should respect language of changed element

hoffmanjoshua opened the pull request:
#### 4f7a7f65db7a33c5fb5304def9ca231fa15ca915
<pre>
AX: live region announcements should respect language of changed element
<a href="https://bugs.webkit.org/show_bug.cgi?id=303321">https://bugs.webkit.org/show_bug.cgi?id=303321</a>
<a href="https://rdar.apple.com/100275523">rdar://100275523</a>

Reviewed by Tyler Wilcock.

Live regions are tied to elements that may have defined languages. If a live region announcement
is posted for one of these elements, the AT should speak this announcement in the right
voice for that language.

To make this possible, we now post announcement notifications on all platforms as attributed
strings (previously, this was limited to iOS). In doing so, we can supply the language
attribute for the AT to consume.

Tests: accessibility/ios-simulator/live-regions/live-region-languages.html
       accessibility/mac/live-regions/live-region-languages.html

* LayoutTests/accessibility/ios-simulator/live-regions/live-region-languages-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/live-regions/live-region-languages.html: Added.
* LayoutTests/accessibility/ios-simulator/live-regions/live-region-types-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-alert-added-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-hidden-content-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-languages-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-languages.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-types-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic-expected.txt:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::computeChanges const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
(WebCore::AXLiveRegionManager::postAnnouncementForChange):
* Source/WebCore/accessibility/AXLiveRegionManager.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::postLiveRegionNotification):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::relayLiveRegionNotification):

Canonical link: <a href="https://commits.webkit.org/303732@main">https://commits.webkit.org/303732@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/beae535b0a26c037400c666cefcd752f8319e989

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133439 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5940 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140995 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135309 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5805 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102073 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136386 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4570 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119606 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82869 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4447 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2045 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113556 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143641 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5610 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38303 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110449 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5692 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4809 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110632 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28043 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4308 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115865 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59360 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5665 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34188 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5511 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69117 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5754 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5621 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -280,29 +309,29 @@ String AXLiveRegionManager::computeAnnouncement(const LiveRegionSnapshot& newSna
+        }
+
+        String removalPrefix = AXRemovedText();
+        stringBuilder.append(removalPrefix);
+        characterCount += removalPrefix.length();
```
Can we move the string rather than copy? We can store the length in a local before moving it.

hoffmanjoshua replied:
Fixed!

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-languages.html:
```diff
@@ -0,0 +1,45 @@
+    accessibilityController.addNotificationListener(notificationCallback);
+    accessibilityController.accessibleElementById("polite-region");
+
+    document.getElementById('spanish-region').textContent = "Manzana";
```
I think we use double quotes wherever possible, i.e. "spanish-region"

twilco commented on LayoutTests/accessibility/ios-simulator/live-regions/live-region-languages.html:
```diff
@@ -0,0 +1,44 @@
+    accessibilityController.addNotificationListener(notificationCallback);
+    accessibilityController.accessibleElementById("region");
+
+    document.getElementById('spanish-region').textContent = "Manzana";
```
double quotes here too

twilco commented on LayoutTests/accessibility/ios-simulator/live-regions/live-region-languages.html:
```diff
@@ -0,0 +1,44 @@
+function notificationCallback(object, notification, userInfo) {
+    if (notification == "AXAnnouncementRequested") {
+        output += `Announcement received, notification: ${notification}, attributed string: ${userInfo["AXAnnouncementRequested"]}\n`;
+        output += `\n`;
```
Can use double quotes instead of format string quotes since we aren't formatting anything

twilco replied:
Also can we just include another \n on the previous line?

hoffmanjoshua replied:
Moved all into one line!

twilco commented on LayoutTests/accessibility/mac/live-regions/live-region-languages.html:
```diff
@@ -0,0 +1,45 @@
+    if (notification == "AXAnnouncementRequested") {
+        output += `Received announcement request:\n`;
+        output += formatAnnouncementUserInfo(userInfo);
+        output += `\n`;
```
Can this be one line using a format string?

hoffmanjoshua replied:
fixed!

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -311,7 +340,8 @@ String AXLiveRegionManager::computeAnnouncement(const LiveRegionSnapshot& newSna
     }
 
-    return makeStringByJoining(strings, " "_s);
+    auto string = stringBuilder.toString();
+    return AttributedString { WTFMove(string), WTFMove(attributes), std::nullopt };
```
Is the local string variable necessary?

hoffmanjoshua replied:
Switched to using `isolatedCopy` to avoid this

twilco replied:
Oh I'm so sorry, I assumed `StringBuilder::toString()` returned a string copy, so now I understand why you had the local. `isolatedCopy` doesn't feel exactly right since that says you're wanting to use the string on a different thread, which I don't think is accurate here? Your original code with a copy + WTFMove may be best, sorry 🤦

twilco reviewed (approved):
r+ with isolatedCopy() issue addressed and passing EWS

hoffmanjoshua committed:
AX: live region announcements should respect language of changed element
https://bugs.webkit.org/show_bug.cgi?id=303321
rdar://100275523

Reviewed by Tyler Wilcock.

Live regions are tied to elements that may have defined languages. If a live region announcement
is posted for one of these elements, the AT should speak this announcement in the right
voice for that language.

To make this possible, we now post announcement notifications on all platforms as attributed
strings (previously, this was limited to iOS). In doing so, we can supply the language
attribute for the AT to consume.

Tests: accessibility/ios-simulator/live-regions/live-region-languages.html
       accessibility/mac/live-regions/live-region-languages.html

* LayoutTests/accessibility/ios-simulator/live-regions/live-region-languages-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/live-regions/live-region-languages.html: Added.
* LayoutTests/accessibility/ios-simulator/live-regions/live-region-types-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-alert-added-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-alert-on-initial-load-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-dynamically-added-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-hidden-content-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-languages-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-languages.html: Added.
* LayoutTests/accessibility/mac/live-regions/live-region-removals-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-types-expected.txt:
* LayoutTests/accessibility/mac/live-regions/live-region-with-atomic-expected.txt:
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::computeChanges const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
(WebCore::AXLiveRegionManager::postAnnouncementForChange):
* Source/WebCore/accessibility/AXLiveRegionManager.h:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::postLiveRegionNotification):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::postPlatformLiveRegionNotification):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::relayLiveRegionNotification):

Canonical link: https://commits.webkit.org/303732@main
